common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,311 @@
+import com.yubico.yubikit.piv.PivSession;
+import com.yubico.yubikit.piv.Slot;
 
+import java.io.IOException;
+import java.io.Serializable;
 import java.security.Principal;
 import java.security.PrivateKey;
 import java.security.cert.X509Certificate;
+import java.util.ArrayList;
 
 //Handles Certificate Based Authentication by means of certificates provisioned onto YubiKeys or the devices themselves.
-//Note that CBA requires API >= 21 (YubiKit SDK min API = 19; ClientCertRequest class available with API > 21)
+//Note that CBA requires API >= 21 (YubiKit SDK min API = 19; ClientCertRequest class available with API >= 21)
 public final class ClientCertAuthChallengeHandler implements IChallengeHandler<ClientCertRequest, Void> {
     private static final String TAG = ClientCertAuthChallengeHandler.class.getSimpleName();
     private static final String ACCEPTABLE_ISSUER = "CN=MS-Organization-Access";
     private Activity mActivity;
     private final YubiKitManager mYubiKitManager;
+    private UsbYubiKeyDevice mDevice;
+
+    //creating nested class to hold details of a Certificate needed for the picker,
+    // including subject, issuer, and slot.
+    public class YubiKitCertDetails implements Serializable {
+        private final String issuerText;
+        private final String subjectText;
+        private final Slot slot;
+
+        public YubiKitCertDetails(String issuerText, String subjectText, Slot slot) {
+            this.issuerText = issuerText;
+            this.subjectText = subjectText;
+            this.slot = slot;
+        }
+
+        public String getIssuerText() {
+            return issuerText;
+        }
+
+        public String getSubjectText() {
+            return subjectText;
+        }
+
+        public Slot getSlot() {
+            return slot;
+        }
+    }
 
     public ClientCertAuthChallengeHandler(@NonNull final Activity activity) {
         mActivity = activity;
-        //Create and start YubiKitManager
+        //set mDevice to null
+        mDevice = null;
+        //Create and start YubiKitManager for UsbDiscovery mode.
+        //When in Usb Discovery mode, Yubikeys that plug into the device will be accessible
+        // once the user provides permission via the Android permission dialog.
         mYubiKitManager = new YubiKitManager(mActivity.getApplicationContext());
         mYubiKitManager.startUsbDiscovery(new UsbConfiguration(), new Callback<UsbYubiKeyDevice>() {
             @Override
             public void invoke(UsbYubiKeyDevice device) {
                 Logger.info(TAG, "A YubiKey device was connected");
-                device.setOnClosed(new Runnable() {
+                //set device
+                mDevice = device;
+                mDevice.setOnClosed(new Runnable() {
                     @Override
                     public void run() {
                         Logger.info(TAG, "A YubiKey device was disconnected");
+                        mDevice = null;
```

Comment:
What's the behavior if the customer unplugs the device in the middle of the flow (let's say they are picking the cert)

is it  
1. We cancel everything - display an error dialog, then cancel the request.
or
2. We let them proceed anyway, then after they click "Continue", then we display the error message and cancel the request - if the device does NOT match.
    > If this is the case, do we refresh the page if a different device is plugged in during the flow?

Reply:
The current implementation is more like number 2, although inconsistent with error handling. I agree though that number 1 is simpler, and I like that it encourages users to keep their smartcard plugged in for the entire authentication process. 
I'll ask Vimala regarding an "unplugged device" dialog... although I think it could potentially not be necessary if the eSTS error message can give more specific info to the user as to how to properly plug in their smartcard (I'll check to see if this is still happening by GA)

Reply:
Alternatively, you could move the cert loader logic into the cert picker dialog, and refreshes upon plugging/unplugging.

Reply:
Messaged Vimala and she said the new server error message should take care of the early unplugging situation (no extra dialog). 
The cert picker refreshing is an interesting idea. This would mean that in the unplugging callback, the current dialog would need to be distinguished so that the current dialog isn't dismissed if it's the cert picker. And there needs to be some similar logic for the plugin callback, too. 
I think I'm still leaning towards just canceling the flow if unplugged, partially because I don't think it will be too difficult for users to just click the back button and retype their upn to restart...  what do you think?

Reply:
[ClientCertRequest](https://developer.android.com/reference/android/webkit/ClientCertRequest#proceed(java.security.PrivateKey,%20java.security.cert.X509Certificate[])): There are three methods that can be used to deal with the request: cancel, ignore, and proceed. Proceed is the only method that takes in parameters (a PrivateKey and a Certificate array). So there aren't a ton of options when it comes to sending the server details as to why the flow is being cancelled.
^As mentioned in another comment, I'll look into creating an error dialog similar to the max attempts one.

Reply:
Adding a TODO comment and would like to get input from the team on the importance of having a dialog pop up when yubikey is unplugged too early (and what the dialog should say).